crates/forge/src/lib.rs

```diff
@@ -421,67 +421,23 @@ fn choose_test_strategy_and_run(
             runner,
             runner_config,
             runner_params,
-            cancellation_tokens,
+            send.clone(),
             send_shut_down.clone(),
         )
     }
 }
 
-fn run_single_test(
-    case: Arc<TestCaseRunnable>,
-    runner: Arc<SierraCasmRunner>,
-    runner_config: Arc<RunnerConfig>,
-    runner_params: Arc<RunnerParams>,
-    cancellation_tokens: Arc<CancellationTokens>,
-    send: Sender<()>,
-    send_shut_down: Sender<()>,
-) -> JoinHandle<Result<TestCaseSummary>> {
-    let exit_first = runner_config.exit_first;
-    tokio::task::spawn(async move {
-        tokio::select! {
-            () = cancellation_tokens.exit_first.cancelled() => {
-                // Stop executing all tests because flag --exit-first'
-                // has been set and one test FAIL
-                Ok(TestCaseSummary::Skipped {})
-            },
-            () = cancellation_tokens.error.cancelled() => {
-                // Stop executing all tests because
-                // one of a test returns Err
-                Ok(TestCaseSummary::Skipped {})
-            },
-
-            result = blocking_run_from_test(vec![], case.clone(),runner,  runner_config.clone(), runner_params.clone(), send.clone(), send_shut_down.clone() ) => {
-                match result? {
-                    Ok(result) => {
-                        if exit_first {
-                            if let TestCaseSummary::Failed { .. } = &result {
-                                cancellation_tokens.exit_first.cancel();
-                            }
-                        }
-                        Ok(result)
-                    }
-                    Err(e) => {
-                        cancellation_tokens.error.cancel();
-                        Err(e)
-                    }
-                }
-            }
-        }
-    })
-}
-
 fn run_with_fuzzing(
     args: Vec<ConcreteTypeId>,
     case: Arc<TestCaseRunnable>,
     runner: Arc<SierraCasmRunner>,
     runner_config: Arc<RunnerConfig>,
     runner_params: Arc<RunnerParams>,
-    cancellation_tokens: Arc<CancellationTokens>,
+    send: Sender<()>,
     send_shut_down: Sender<()>,
 ) -> JoinHandle<Result<TestCaseSummary>> {
     tokio::task::spawn(async move {
-        let cancellation_fuzzing_token = CancellationToken::new();
-        let (send, mut rec) = channel(1);
+        let (fuzzing_send, mut fuzzing_rec) = channel(1);
```

Comment:
I think we could check if `send.is_closed()` as the beginning of this function. Example: we have 100 fuzzing test functions, each running 1000 tests, all in `tests` folder (which means one crate). We use `--exit-first`, the first test fails. Then we have to wait for the 99 * 1000 tests to skip instead of returning `Skipped` here, which will happen 99 times instead of 99000 times. There is prob an exaggeration when it comes to number, it is about the idea. WDYT? cc @cptartur

Reply:
I added